```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r make_seurat1}
pbmc.data <- Read10X(data.dir ="C:/Users/salwa/OneDrive/Desktop/summer_projsct_scRNA_seq/first tutorial/hg19/")
```
This line of code reads the scRNA-seq data from the directory
The Read10X() function is commonly used to read data in the 10x Genomics format, which is a common file format for scRNA-seq data. 

```{r make_seurat2}
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
```
I got a warning part that says feature names cannot have underscore and it will replace with dashes 
that means the file we put in R studio has _ and the R changed in to a – and it won’t affect us on any way 
we create a new variable:
test<-rownames(pbmc.data)
then 
test3<-test[grep(“_”, test)]
this will show us all the words or numbers that have what in between those “” in this case it will show us all the row names(genes) that have an _
to check it write
test3
and these are all the genes that had _ and the R studio change it in to -
This line initializes a Seurat object called pbmc using the CreateSeuratObject() function. 
The count matrix is typically a sparse matrix where rows represent genes and columns represent individual cells, with each element representing the number of RNA molecules detected for a given gene in a specific cell.
project: The project parameter is optional and is used to assign a project name to the Seurat object. It is a string that helps when working with multiple datasets or experiments, allowing you to easily identify and distinguish between different projects. In this case, the project name "pbmc3k" is assigned to the pbmc Seurat object.
min.cells( removes genes) :this is set to 3, which means that any cell with fewer than 3 detected molecules  will be considered low quality or potentially unreliable and will be excluded from downstream analysis. This filtering step helps remove cells that may have low-quality data or low capture efficiency, improving the overall quality of the dataset. 


I got a warning part that says feature names cannot have underscore and it will replace with dashes 
that means the file we put in R studio has _ and the R changed in to a – and it won’t affect us on any way 
```{r make_seurat3}
test<-rownames(pbmc.data)
```
then 
```{r make_seurat4}
test3<-test[grep(“_”, test)]
```
this will show us all the words or numbers that have what in between those “” in this case it will show us all the row names(genes) that have an _
to check it write


```{r make_seurat5}
test3
```

and these are all the genes that had _ and the R studio change it in to -

This line initializes a Seurat object called pbmc using the CreateSeuratObject() function. 
The count matrix is typically a sparse matrix where rows represent genes and columns represent individual cells, with each element representing the number of RNA molecules detected for a given gene in a specific cell.
project: The project parameter is optional and is used to assign a project name to the Seurat object. It is a string that helps when working with multiple datasets or experiments, allowing you to easily identify and distinguish between different projects. In this case, the project name "pbmc3k" is assigned to the pbmc Seurat object.
min.cells( removes genes) :this is set to 3, which means that any cell with fewer than 3 detected molecules  will be considered low quality or potentially unreliable and will be excluded from downstream analysis. This filtering step helps remove cells that may have low-quality data or low capture efficiency, improving the overall quality of the dataset. 



This line displays the information about the pbmc Seurat object, which provides a summary of the dataset and its preprocessing steps. It shows details such as the number of cells, features, and the steps performed during preprocessing.
```{r make_seurat}
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
```
This code calculates and assigns the percentage of mitochondrial genes to a new column named "percent.mt" in the pbmc object.
 The PercentageFeatureSet function is a custom function that calculates the percentage of genes matching a specified pattern, which in this case is genes starting with "MT-". The function takes the pbmc object and the pattern as arguments.
```{r make_seurat6}
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```
VlnPlot(): This is a function in Seurat used to generate violin plots. 
Violin plots are useful for visualizing the distribution of data across different groups or categories.
pbmc: This is the Seurat object on which the violin plot will be created. 
features = c("nFeature_RNA", "nCount_RNA", "percent.mt"): This argument specifies the features (variables) to be included in the violin plot. In this case, the plot will display the distributions of three features:
"nFeature_RNA": This represents the number of detected features (genes) for each cell in the dataset. It provides information about the complexity or richness of gene expression in each cell.
"nCount_RNA": This represents the total RNA molecule count for each cell. It reflects the overall expression level or total transcriptional activity of each cell.
"percent.mt": This is the percentage of mitochondrial genes in each cell, as calculated in a previous step of the analysis. It provides insight into potential mitochondrial content or stress in the cells.
By specifying these features, the violin plot will show the distributions of these three variables across the cells in the pbmc dataset.
ncol = 3: sets the number of columns in the resulting plot. In this case, it specifies that the violin plots should be arranged in three columns, corresponding to the three specified features.
```{r make_seurat7}
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
```
plot1: it creates a new object called plot1 to store the scatter plot.
FeatureScatter(): is used to generate a scatter plot of two features. 
feature1 = "nCount_RNA": specifies the first feature to be plotted on the x-axis of the scatter plot. In this case, the feature is "nCount_RNA", which represents the total RNA molecule count for each cell.
feature2 = "percent.mt": specifies the second feature to be plotted on the y-axis of the scatter plot. In this case, the feature is "percent.mt", which represents the percentage of mitochondrial genes in each cell.
By specifying these two features, the FeatureScatter function will generate a scatter plot showing the relationship or correlation between the total RNA molecule count and the percentage of mitochondrial genes for each cell in the pbmc dataset.
The resulting scatter plot, stored in the plot1 object.
```{r make_seurat8}
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```
plot2: This line creates a new object called plot2 to store the scatter plot.
FeatureScatter(): This function is used to generate a scatter plot of two features. 
pbmc: This is the Seurat object that contains the scRNA-seq data on which the scatter plot will be created. It represents the processed and analyzed data.
feature1 = "nCount_RNA": specifies the first feature to be plotted on the x-axis of the scatter plot. In this case, the feature is "nCount_RNA", which represents the total RNA molecule count for each cell.
feature2 = "nFeature_RNA": specifies the second feature to be plotted on the y-axis of the scatter plot. In this case, the feature is "nFeature_RNA", which represents the number of detected features (genes) for each cell.
By specifying these two features, the FeatureScatter function will generate a scatter plot showing the relationship or correlation between the total RNA molecule count and the number of detected features for each cell in the pbmc dataset.
It helps to understand the relationship between the overall expression level (total RNA molecule count) and the complexity of gene expression (number of detected features) across the cells in the dataset.
```{r make_seurat9}
plot1 + plot2
```
The + operator is used to overlay the two plots together in a single figure. we will obtain a combined figure that includes the contents of both plot1 and plot2. The resulting plot will show the scatter plot generated by plot1, representing the relationship between the total RNA molecule count and the percentage of mitochondrial genes, as well as the scatter plot generated by plot2, representing the relationship between the total RNA molecule count and the number of detected features.
 The resulting figure will provide insights into both the correlation between total RNA molecule count and mitochondrial genes and the correlation between total RNA molecule count and the complexity of gene expression.
```{r make_seurat10}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```
nFeature_RNA > 200: This filters the cells in pbmc based on the number of detected features. It selects cells where the number of detected features is greater than 200.
nFeature_RNA < 2500: This filters the cells based on the number of detected features, selecting cells where the number of detected features is less than 2500.
percent.mt < 5: This filters the cells based on the percentage of mitochondrial genes. It selects cells where the percentage of mitochondrial genes is less than 5.
By combining these, it ensures that only cells fulfilling all three criteria are retained in the filtered pbmc object.
```{r make_seurat11}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
```
NormalizeData(): This is used to normalize the expression data
scale.factor = 10000: This specifies the scale factor to be applied during the normalization process. 
The scale factor determines the desired median library size for the normalized data. In this case, a scale factor of 10000 is chosen.
the pbmc object will undergo log-normalization and scaling. 
The expression data within the object will be transformed, ensuring that the resulting data has a more standardized and comparable scale across cells.
This normalization step helps to remove biases caused by differences in library sizes and improves the accuracy and reliability of downstream analyses such as clustering, differential expression analysis, and visualization.
```{r make_seurat12}
pbmc <- NormalizeData(pbmc)
```
By executing NormalizeData(pbmc), the pbmc  will undergo normalization using the default normalization method. The specific normalization method used by default in Seurat is called "LogNormalize." This method involves taking the logarithm of the raw counts and then scaling them to achieve a more balanced distribution of gene expression values across cells.
```{r make_seurat13}
{pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)}
```
FindVariableFeatures(): This  is used to identify features that exhibit high variability across cells. These features are often informative for downstream analysis, as they capture the most biologically relevant differences between cells.
selection.method = "vst": This specifies the method to be used for feature selection. VST is a popular method for estimating the biological variability of scRNA-seq data.
nfeatures = 2000: This specifies the number of variable features to be identified. In this case, 2000 features will be selected as the most variable features.
By using this code: the pbmc object will be updated to include information about the variable features. These features represent genes that exhibit the highest variability across cells, capturing important biological differences.
Identifying variable features is a critical step in scRNA-seq data analysis as it helps to focus on genes that contribute the most to cellular heterogeneity and biological diversity. 
```{r make_seurat14}
top10 <- head(VariableFeatures(pbmc), 10)
```
head(VariableFeatures(pbmc), 10): This uses the head function to select the top 10 variable features from the list obtained using VariableFeatures. 
It stores the top 10 variable features in the top10 object.
```{r make_seurat15}
plot1 <- VariableFeaturePlot(pbmc)
```
This creates a new object called plot1 to store the plot generated by VariableFeaturePlot(pbmc).
```{r make_seurat16}
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
```
This is used to label specific points on a plot. In this case, it is used to label the top 10 variable features (stored in the top10 object) on the plot generated by VariableFeaturePlot.
plot2: This line of code creates a new object called plot2 to store the plot with labeled points.
```{r make_seurat17}
plot1 + plot2
```
This code combines plot1 and plot2 by using the +. we obtain a combined figure that includes the variability plot generated by VariableFeaturePlot(pbmc) and the labeled points corresponding to the top 10 variable features. The resulting plot provides visual insights into the distribution of variable features a
```{r make_seurat18}
all.genes <- rownames(pbmc)
```
It retrieves the names of all the genes present in the pbmc object and stores them in the all.genes variable. 
The rownames() function is used to extract the row names of the expression matrix in the pbmc object, which correspond to the gene names.
```{r make_seurat19}
pbmc <- ScaleData(pbmc, features = all.genes)
```
the expression data in the pbmc object will be transformed using scaling techniques.
 Scaling adjusts the expression values of each gene across cells to have similar distributions or scales. It helps to remove the influence of gene-specific differences in expression ranges, allowing for better comparison and interpretation of gene expression patterns across cells.
The resulting pbmc object will contain the scaled expression data, which can then be used for downstream analyses such as clustering.
```{r make_seurat20}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```
RunPCA() is used to run principal component analysis (PCA) on data. 
features = VariableFeatures(object = pbmc): This  specifies the features  to be used for PCA. In this case, VariableFeatures() is called to retrieve the previously identified variable features from the pbmc object. These features capture the genes that exhibit the highest variability across the cells.
after using this code the pbmc object will be updated to include the results of the PCA. The expression data within the object will undergo PCA using the selected features.
```{r make_seurat21}
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)
```
pbmc[["pca"]]: This accesses the PCA results within the pbmc object.
The pca is a named list or matrix that contains the calculated principal components and associated metadata.
dims = 1:5: This specifies the dimensions (principal components) to be displayed in the output. In this case, it selects the first five dimensions of the PCA results.
nfeatures = 5: This specifies the number of features to be displayed for each dimension. 
It restricts the output to show information about the top five features contributing to each selected dimension.
the code displays information about the PCA results for the specified dimensions and associated top features.
The output will typically include the following information for each selected dimension:
The name or index of the dimension 
The percentage of the total variance explained by that dimension.The cumulative variance explained up to that dimension.
Information about the top contributing features, such as their names or indices and the magnitude of their contributions to that dimension.
```{r make_seurat22}
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
```
VizDimLoadings(): This is used to visualize the loadings of the principal components (PCs) in a dimensionality reduction technique. 
dims = 1:2: This specifies the dimensions (PCs) to be visualized. the first two dimensions (PC1 and PC2) are selected.
reduction = "pca": This specifies the dimensionality reduction technique used. In this case, it is set to "pca" to indicate that the loadings of the principal components in PCA should be visualized.
 the function generates a plot that visualizes the loadings of the selected principal components in the PCA results of the pbmc object.
```{r make_seura23t}
DimPlot(pbmc, reduction = "pca")
```
DimPlot(): This is used to create plots.
reduction = "pca": This specifies the dimensionality reduction technique used. In this case
So this function generates a plot that visualizes the cells in the reduced dimensional space obtained through PCA.
The specific type of plot generated by DimPlot depends on the dimensionality reduction technique used. For PCA, the resulting plot is typically a scatterplot where each point represents a cell, and the position of the points in the plot is determined by their scores on the selected principal components (PCs). The color or shape of the points can be used to represent different metadata information, such as cell type or experimental condition.
This plot is useful for visualizing the distribution of cells in the reduced dimensional space and identifying clusters or groups of cells that exhibit similar gene expression profiles. It allows for the exploration of the relationships and structure in the dataset and can aid in the identification of biologically meaningful patterns and differences among cells.
```{r make_seurat24}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
```
This creates a heatmap based on the results of a dimensionality reduction technique, specifically using the first dimension (PC1) in this case. Let's break down the code:
DimHeatmap(): This is used to create a heatmap 
dims = 1: This  specifies the dimension (PC1) to be used for creating the heatmap.
cells = 500: This specifies the number of cells to include in the heatmap. 
balanced = TRUE: This specifies whether to balance the number of cells included from each cluster/group. When set to TRUE, the function attempts to include a balanced number of cells from each cluster/group in the heatmap.
```{r make_seurat25}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```
DimHeatmap(): This is used to create heatmaps based on the results of dimensionality reduction
dims = 1:15: This specifies the dimensions (PCs) to be used for creating the heatmaps. We selected PC1 to PC15, representing the first 15 dimensions.
cells = 500: This parameter specifies the number of cells to include in each heatmap. In this case, it selects 500 cells based on their scores.
balanced = TRUE: This specifies whether to balance the number of cells included from each cluster/group in each heatmap. When set to TRUE, the function attempts to include a balanced number of cells from each cluster/group in each heatmap.
This  generates a series of heatmaps, each visualizing the gene expression patterns of the selected cells based on their scores in the corresponding dimension (PC1 to PC15).
Each heatmap in the series will typically have cells as rows and genes as columns. The color intensity in each heatmap represents the gene expression values, allowing for the visualization of gene expression patterns across the selected cells in each dimension.
```{r make_seurat26}
pbmc <- JackStraw(pbmc, num.replicate = 100)
```
JackStraw(): This  is used to perform a jackstraw analysis, It assesses the statistical significance of each principal component (PC) in a dimensionality reduction technique, such as PCA. 
The jackstraw analysis estimates the p-values for each PC, indicating the likelihood of observing the observed variation in the data by chance.
num.replicate = 100: This specifies the number of replicate permutations to be performed in the jackstraw analysis. In this case, 100 permutations are used to estimate the p-values.
This code performs the jackstraw analysis on the principal components obtained from PCA in the pbmc object.
The output of the JackStraw function  includes:
A plot showing the ordered jackstraw scores for each PC, which helps identify the PCs with significant or non-random patterns of variation.
A plot showing the ordered observed eigenvalues and the 95% confidence intervals obtained from the jackstraw analysis, which assists in identifying the significant PCs.
A table or list containing the p-values for each PC, indicating the statistical significance of their observed variation.
```{r make_seurat27}
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
```
ScoreJackStraw(): This is used to calculate the scores for each PC based on the results of the jackstraw analysis. 
dims = 1:20: This specifies the dimensions (PCs) for which we want to calculate the scores. In this case, we select PC1 to PC20.
this code calculates the scores for each selected PC based on the jackstraw analysis results in the pbmc object.
The resulting scores are typically stored in the pbmc object, allowing us to access and analyze them as needed. The scores provide information about the strength or significance of each PC in capturing the observed variation in the scRNA-seq data. Higher scores indicate PCs that capture more of the meaningful variation, while lower scores suggest PCs with random or non-significant patterns of variation.
```{r make_seurat28}
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
```
ScoreJackStraw(): This function is used to calculate the scores for each PC based on the results of the jackstraw analysis. The scores represent the strength or significance of each PC in capturing the observed variation in the data.
dims = 1:20: This  specifies the dimensions (PCs) for which you want to calculate the scores. In this case, PC1 to PC20.
```{r make_seurat29}
JackStrawPlot(pbmc, dims = 1:15)
```
JackStrawPlot(): This generates a plot to visualize the results of the jackstraw analysis.
dims = 1:15: This specifies the dimensions (PCs) to be included in the plot. In this case, it selects PC1 to this function creates a plot that displays the results of the jackstraw analysis for the selected principal components.
The resulting plot consists of two main components:
Jackstraw plot: This plot visualizes the ordered jackstraw scores for each principal component. The jackstraw scores represent the strength or significance of each PC in capturing the observed variation. The plot allows us to identify PCs with significant or non-random patterns of variation.
```{r make_seurat30}
ElbowPlot(pbmc)
```
ElbowPlot(): This function generates an elbow plot to visualize the explained variance for different numbers of PCs.
 this function creates a plot that shows the explained variance as a function of the number of PCs.
The resulting plot typically consists of two main components:
The x-axis represents the number of PCs included in the analysis.
The y-axis represents the explained variance, which indicates the proportion of total variation in the data captured by each number of PCs.
The plot is called an "elbow plot" because it often exhibits a distinctive bend, or "elbow," where the explained variance starts to level off. This bend indicates the optimal number of PCs to retain for further analysis. The optimal number is often chosen at the point where adding more PCs does not significantly increase the explained variance.
```{r make_seurat31}
pbmc <- FindNeighbors(pbmc, dims = 1:10)
```
FindNeighbors(): This is used to compute the nearest neighbors for each cell in the dataset based on a distance metric calculated from a dimensionality reduction analysis.
dims = 1:10: This specifies the dimensions (PCs) to be used for finding the neighbors. In this case,we  selectPC1 to PC10.
this code calculates the nearest neighbors for each cell in the dataset based on the selected dimensions.
The output includes an updated pbmc object that contains the computed neighborhood information. This information is stored as a graph or a matrix that represents the connectivity or similarity between cells based on the selected dimensions. Each cell's neighbors are determined by their proximity in the reduced-dimensional space.
```{r make_seurat32}
pbmc <- FindClusters(pbmc, resolution = 0.5)
```
FindNeighbors(): This is used to compute the nearest neighbors for each cell in the dataset based on a distance metric calculated from a dimensionality reduction analysis.
dims = 1:10: This specifies the dimensions (PCs) to be used for finding the neighbors. In this case,we  selectPC1 to PC10.
this code calculates the nearest neighbors for each cell in the dataset based on the selected dimensions.
The output includes an updated pbmc object that contains the computed neighborhood information. This information is stored as a graph or a matrix that represents the connectivity or similarity between cells based on the selected dimensions. Each cell's neighbors are determined by their proximity in the reduced-dimensional space.
```{r make_seurat33}
pbmc <- FindClusters(pbmc, resolution = 0.5)
```
FindClusters(): This is used to identify clusters or groups of cells based on their transcriptional similarity.
resolution = 0.5: This specifies the resolution parameter, which controls the granularity of the clustering. A higher resolution value leads to more fine-grained clusters, while a lower value produces fewer and larger clusters.
The output of the FindClusters function is typically an updated pbmc object that includes the assigned cluster labels for each cell. Each cell is assigned to a specific cluster based on its transcriptional similarity to other cells.

```{r make_seurat34}
head(Idents(pbmc), 5)
```

```{r make_seurat35}
pbmc <- RunUMAP(pbmc, dims = 1:10)
```
RunUMAP(): This is used to perform dimensionality reduction using the UMAP algorithm.
dims = 1:10: This specifies the dimensions (PCs) to be used for the UMAP computation. In this case, we select PC1 to PC10.
this computes the UMAP representation for the cells in the pbmc object based on the selected dimensions.

```{r make_seurat36}
DimPlot(pbmc, reduction = "umap")
```
DimPlot():  is used to create plots to visualize cells in a dimensionality-reduced space.
reduction = "umap": This specifies the type of dimensionality reduction to be plotted. We select the UMAP coordinates as the basis for visualization.
The resulting plot displays the cells as points or dots in a two-dimensional space. Each point represents a cell, and its position reflects the UMAP coordinates calculated for that cell. Cells with similar transcriptional profiles are projected closer to each other in the UMAP space, allowing for the identification of clusters or groups of cells with shared characteristics.



```{saveRDS(pbmc, file = "C:/Users/salwa/OneDrive/Desktop/summer_projsct_scRNA_seq/first tutorial/hg19/")
```
saveRDS(): This  is used to save an R object to a file in RDS format.
```{r make_seurat37}
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2, min.pct = 0.25)
```
FindMarkers(): This is used to identify differentially expressed genes that serve as markers for specific cell clusters or groups.
ident.1 = 2: This specifies the cluster ID for which we want to find markers. 
min.pct = 0.25: This parameter specifies the minimum percentage of cells in a cluster that should express a gene to be considered as a marker. In this case, it requires that a gene is expressed in at least 25% of cells in cluster 2.
The output of the FindMarkers function includes a data frame that contains information about the differentially expressed genes, such as gene names, log fold changes, p-values, and adjusted p-values. The markers are genes that exhibit significant differential expression between the specified cluster and the other clusters in the dataset.

```{r make_seurat38}
head(cluster2.markers, n = 5)
```
The head() function is used to display the first few rows of a data frame or matrix in R. here we want to display the first five rows of the cluster2.markers object, which is the result of the FindMarkers function.
```{r make_seurat39}
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```
FindAllMarkers(): This function is used to identify differentially expressed genes that serve as markers for each cluster in a Seurat object.
only.pos = TRUE: This means that only genes that are positively enriched in a cluster will be considered as markers. Genes that are downregulated or not differentially expressed will be excluded.
min.pct = 0.25: This put the minimum percentage of cells in a cluster that should express a gene to be considered as a marker. Here it requires that a gene is expressed in at least 25% of cells in each cluster.
logfc.threshold = 0.25: This sets a threshold for the log fold change value to filter the markers. Genes with a log fold change greater than 0.25 will be considered as markers.

```{r make_seurat40}
pbmc.markers %>%
```

```{r make_seurat41}
group_by(cluster) %>%
```

```{r make_seurat42}
slice_max(n = 2, order_by = avg_log2FC)
```
%>%: The pipe operator %>% allows you to chain multiple operations together, passing the result of one function as the input to the next function.
group_by(cluster): This function is used to group the data by the "cluster" column in the pbmc.markers object. This step allows subsequent operations to be performed on each cluster separately.
slice_max(n = 2, order_by = avg_log2FC): This is used to subset the data and keep only the top two rows (genes) for each cluster based on the "avg_log2FC" column. It selects the genes with the highest average log fold change within each cluster.
The pbmc.markers object is grouped by the "cluster" column.
The resulting output will be a subset of the pbmc.markers object, containing only the top two genes with the highest average log fold change for each cluster.
This code is useful for identifying the most significant marker genes that exhibit the largest fold changes within each cluster, which can provide insights into the distinguishing gene expression patterns of different cell types or states.

```{r make_seurat43}
cluster0.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
```
FindMarkers(): This function is used to identify differentially expressed genes that serve as markers for specific cell clusters or groups.
ident.1 = 0: This specifies the cluster ID for which we want to find markers. In this case, it selects cluster ID 0.
logfc.threshold = 0.25: This sets a threshold for the log fold change value to filter the markers. Genes with a log fold change greater than 0.25 will be considered as markers.
test.use = "roc": This  specifies the statistical test to be used for finding markers.
 The "roc" option indicates the receiver operating characteristic (ROC) test will be used. 
only.pos = TRUE: This specifies that only genes that are positively enriched in the cluster will be considered as markers. Genes that are downregulated or not differentially expressed will be excluded.
```{r make_seurat44}
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
```
VlnPlot(): This function is used to create violin plots.
features = c("MS4A1", "CD79A"): This specifies the genes ("MS4A1" and "CD79A") for which you want to create violin plots.
By executing VlnPlot(pbmc, features = c("MS4A1", "CD79A")), this will create a violin plot for each specified gene ("MS4A1" and "CD79A") in the pbmc object.
The resulting plot will display violin plots for the expression of "MS4A1" and "CD79A" genes across different cell populations or clusters in the dataset. Violin plots provide information about the distribution of gene expression levels within each cell population, showing the density of expression values at different levels. The width of the violin represents the density, and the shape represents the distribution of expression values.
```{r make_seurat45}
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
```
VlnPlot(): This function is used to create violin plots.
features = c("NKG7", "PF4"): This specifies the genes ("NKG7" and "PF4") for which we want to create violin plots.
slot = "counts": This specifies the data slot in the Seurat object from which to retrieve the expression values. In this case, it specifies the "counts" slot, which contains the raw count data.
log = TRUE: This parameter specifies whether the expression values should be log-transformed before plotting. Setting it to TRUE means that the expression values will be log-transformed.
the function will create violin plots for the "NKG7" and "PF4" genes using the raw count data from the pbmc object. The expression values will be log-transformed before plotting.
The resulting plot will display violin plots for the log-transformed expression levels of "NKG7" and "PF4" genes across different cell populations or clusters in the dataset. Violin plots provide information about the distribution of gene expression levels within each cell population, showing the density of expression values at different levels. The log transformation helps to visualize the gene expression distributions more effectively, especially when the data has a wide dynamic range.

```{r make_seurat46}
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A"))
```
FeaturePlot(): This function is used to create feature plots.
features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A"): This parameter specifies the genes ("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A") for which you want to create feature plots.
it will show expression of selected genes across different cell populations or clusters in the dataset. Each gene will have a separate plot, and each plot will show the expression levels of the gene in different cell populations or clusters. 
```{r make_seurat47}
pbmc.markers %>%
```

```{r make_seurat48}
group_by(cluster) %>%
```

```{r make_seurat49}
top_n(n = 10, wt = avg_log2FC) -> top10
```
%>%: The pipe operator %>% allows us to chain multiple operations together, passing the result of one function as the input to the next function.
group_by(cluster): This function  is used to group the data by the "cluster" column in the pbmc.markers object. This step allows subsequent operations to be performed on each cluster separately.
top_n(n = 10, wt = avg_log2FC): This e is used to select the top 10 rows (genes) for each cluster based on the "avg_log2FC" column. It selects the genes with the highest average log fold change within each cluster.
top10: This assigns the result of the previous operation to the variable top10.
```{r make_seurat50}
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```
DoHeatmap(): This function is used to create a heatmap of gene expression.
features = top10$gene: This parameter specifies the genes to be included in the heatmap. The top10$gene retrieves the gene names from the top10 data frame that was previously defined.
the function will create a heatmap of gene expression in the pbmc object. 

```{r make_seurat51}
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono","NK", "DC", "Platelet")
```
new.cluster.ids: This is the name of the new vector that will store the cluster names.
c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono", "NK", "DC", "Platelet"): This is a character vector containing the names of different cell populations or clusters.
the vector new.cluster.ids will be created and assigned with the provided cluster names.
```{r make_seurat52}
names(new.cluster.ids) <- levels(pbmc)
```
names(new.cluster.ids): This is accessing the names attribute of the new.cluster.ids vector.
 levels(pbmc): This is assigning the levels of the cluster identity variable in the pbmc Seurat object to the names of the new.cluster.ids vector.
The names of the new.cluster.ids vector will be replaced with the levels of the cluster identity variable in the pbmc object. This ensures that the names in new.cluster.ids align with the existing cluster identities in the pbmc object.
This type of assignment is useful when you want to update or rename the cluster identities in a Seurat object
```{r make_seurat53}
pbmc <- RenameIdents(pbmc, new.cluster.ids)
```
RenameIdents(): This  is used to update or rename the cluster identities in a Seurat object.
new.cluster.ids: This is the vector that contains the new names for the cluster identities.
this function will update the cluster identities in the pbmc object with the names from the new.cluster.ids vector. Each cluster in the pbmc object will be assigned the corresponding name from the new.cluster.ids vector.
This operation is useful when you want to change the cluster names in a Seurat object to more meaningful or informative names based on your specific context or analysis. It helps in improving the interpretability and clarity of the cluster labels in downstream analyses and visualizations.
```{r make_seurat54}
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```
DimPlot(): This function is used to create a plot of the cells in a dimensionality-reduced space.
reduction = "umap": This parameter specifies the type of dimensionality reduction to be visualized. 
label = TRUE: This parameter specifies that labels should be displayed for the data points in the plot. This is helpful for identifying specific cell populations or clusters.
pt.size = 0.5: This parameter adjusts the size of the data points in the plot. A value of 0.5 indicates a smaller size for the data points.
+ NoLegend(): This adds the NoLegend() function to remove the legend from the plot.
```{r make_seurat55}
saveRDS(pbmc, file = "C:/Users/salwa/OneDrive/Desktop/summer_projsct_scRNA_seq/first tutorial/hg19/")
```

```{r make_seurat56}
pbmc3k.final <- pbmc
```

```{r make_seurat57}
pbmc3k.final$groups <- sample(c("group1", "group2"), size = ncol(pbmc3k.final), replace = TRUE)
```

```{r make_seurat58}
head(pbmc3k.final)
```


```{r make_seurat59}
features <- c("LYZ", "CCL5", "IL32", "PTPRCAP", "FCGR3A", "PF4")
```
creates a vector named features that contains the names of six genes of interest.
The genes listed in the vector are:
LYZ
CCL5
IL32
PTPRCAP
FCGR3A
PF4
By storing these gene names in the features vector, you can easily reference and use them in subsequent analysis steps, such as plotting, differential expression analysis, or subset selection based on specific genes.

```{r make_seurat60}
pbmc3k.final
```

```{r make_seurat61}
RidgePlot(pbmc3k.final, features = features, ncol = 2)
```
RidgePlot(): This function is used to create a ridge plot, also known as a joyplot, which displays the density distribution of multiple variables simultaneously.
features = features: This specifies the features (genes) that will be plotted in the ridge plot. In this case, the features vector that I previously defined is used to specify the genes of interest.
ncol = 2: This parameter determines the number of columns in which the ridge plots will be arranged. In this case, the ridge plots will be arranged in two columns.
By executing the  a ridge plot will be created for the pbmc3k.final object. The ridge plot will show the density distribution of the expression levels of the specified genes (features) across the cells in the dataset. Each gene will have its own ridge plot, and they will be arranged in two columns. This type of visualization is useful for exploring the distribution and variation of gene expression levels within and across different cell populations or clusters.
we look at clusters where the the horizontal line shows the percentage of each cell with different expression 
here we can ask a question by looking at the plot what cell clusters have the most expression or least expression of specific gene in a specific clusters 
sometimes we pick this specific genes because we already know that it will be overexpressed in a certain type of cells to know the type, as we go horizontally it means the expression increase but vertically shows the density of cells at this specific expression

```{r make_seurat62}
VlnPlot(pbmc3k.final, features = features)
```
VlnPlot(): This function is used to create a violin plot, which visualizes the distribution of gene expression across different cell populations.
features = features: This parameter specifies the genes of interest (features) that will be plotted in the violin plot. In this case, the features vector that I previously defined is used to specify the genes.
By executing the code, a violin plot will be generated for each gene specified in the features vector. The violin plot represents the distribution of expression levels for each gene across the cells in the pbmc3k.final object. The width of each violin corresponds to the density of expression levels, while the white dot indicates the median expression. This visualization is useful for comparing the expression patterns of different genes across different cell populations or clusters in your dataset.
it means that I can see an expression level in different cells or clusters and the gene expression is plotted in the shape of violin 
where as the expression of the gene increase the width of the violin increase 
in this case it is int single cell RNA seq
The black dot represent a certain cell expression 
the little tail because of 0 expression 

```{r make_seurat63}
FeaturePlot(pbmc3k.final, features = features)
```
FeaturePlot(): This function is used to create a feature plot, which displays the expression levels of genes of interest in individual cells.
features = features: This parameter specifies the genes of interest (features) that will be plotted in the feature plot. The features vector I provided earlier is used to specify the genes.
By using the code FeaturePlot(pbmc3k.final, features = features), a feature plot will be generated for each gene specified in the features vector. Each plot will display the expression levels of the corresponding gene in individual cells or cell clusters in the pbmc3k.final object. The x-axis corresponds to the cells in the dataset, and the y-axis represents the expression values for the respective gene.
UMAP (Uniform Manifold Approximation and Projection) algorithm. UMAP is a dimensionality reduction technique commonly used in single-cell RNA sequencing (scRNA-seq) analysis.
UMAP is designed to capture the underlying structure and relationships within high-dimensional data and project it into a lower-dimensional space. In the case of scRNA-seq data, UMAP is often used to visualize the cellular heterogeneity and identify distinct cell populations or clusters.
In the feature plot, the UMAP 1 and UMAP 2 dimensions represent the two-dimensional projection of the cells in the dataset based on their gene expression profiles. Each point in the plot corresponds to a single cell, and its position in the UMAP space reflects its similarity to other cells based on gene expression patterns. Cells that are close to each other in the UMAP plot are more similar in terms of gene expression, while cells that are far apart are more dissimilar. The feature plot overlaid on the UMAP visualization allows you to explore the expression levels of specific genes (features) across different cells and their potential associations with cell clusters or subpopulations identified by the UMAP algorithm. This can provide valuable insights into the gene expression patterns and functional differences between distinct cell types
we are going to plot specific genes that we call before a vector (features) and we are going to see it is expression in the individual cells, as the color gets deeper the expression increase, every dot represent a specific a cell 
where they tend to cluster according to the type as tutorial one but instead using colors as a diff type of cell we use the color to check the expression of specific gene



```{r make_seurat64}
DotPlot(pbmc3k.final, features = features) + RotatedAxis()
```
DotPlot(): This function is used to create a dot plot, which visualizes the expression levels of genes in different cell populations.
features = features: This parameter specifies the genes of interest (features) that will be plotted in the dot plot. The features vector you provided earlier is used to specify the genes.
+ RotatedAxis(): This addition applies rotated axis labels in the dot plot to improve the readability of gene names when there are many genes of interest.
By executing the code DotPlot(pbmc3k.final, features = features) + RotatedAxis(), a dot plot will be generated for each gene specified in the features vector. The dot plot represents the expression levels of each gene in different cell populations or clusters. Each dot corresponds to a single cell, and the color or size of the dot represents the expression level of the gene in that cell. The rotated axis labels ensure that the gene names are more easily readable when there are multiple genes plotted. This visualization allows for the comparison of gene expression patterns across different cell populations or clusters, highlighting the relative expression levels of the selected genes.
it measures the expression of each gene in each single cell where the size of the dot represent the percentage that are expressed and the color is the average expressed as the size increase the percentage increase and as th color go deeper average expression increased 
if I have 100 cell I dtect a expression of 5 in each every single plateile it will be a very big dot bur very light color bec the expression is very low 



```{r make_seurat65}
DoHeatmap(subset(pbmc3k.final, downsample = 100), features = features, size = 3)
```
DoHeatmap(): This function is used to create a heatmap, which visualizes the expression levels of genes across cells.
subset(pbmc3k.final, downsample = 100): This subsets the pbmc3k.final Seurat object by randomly sampling 100 cells from the dataset. It reduces the number of cells to be included in the heatmap for better visualization.
features = features: This parameter specifies the genes of interest (features) that will be included in the heatmap. The features vector you provided earlier is used to specify the genes.
size = 3: This parameter sets the cell size of the heatmap to 3, controlling the resolution of the heatmap cells.
By executing the code a heatmap will be generated for the specified subset of cells. The heatmap will display the expression levels of the selected genes (features) across the sampled cells. Each row represents a gene, each column represents a cell, and the color of each cell in the heatmap corresponds to the expression level of the gene in that cell. The heatmap allows for the identification of gene expression patterns and clustering of cells based on their expression profiles.


```{r make_seurat66}
FeaturePlot(pbmc3k.final, features = "MS4A1")
```
FeaturePlot(): This function is used to create a feature plot, which displays the expression levels of genes of interest in individual cells or cell clusters.
features = "MS4A1": This parameter specifies the gene of interest that will be plotted in the feature plot. In this case, the gene "MS4A1" is selected.
By executing the code a feature plot will be generated specifically for the "MS4A1" gene. The plot will display the expression levels of the "MS4A1" gene in individual cells or cell clusters in the pbmc3k.final object. This visualization helps in examining the expression pattern and heterogeneity of the "MS4A1" gene across different cells or cell populations. It allows for the identification of cell types or states that exhibit high or low expression levels of the "MS4A1" gene.

```{r make_seurat67}
FeaturePlot(pbmc3k.final, features = "MS4A1", min.cutoff = 1, max.cutoff = 3)
```
FeaturePlot(): This function is used to create a feature plot, which displays the expression levels of genes of interest in individual cells or cell clusters.
features = "MS4A1": This parameter specifies the gene of interest that will be plotted in the feature plot. In this case, the gene "MS4A1" is selected.
min.cutoff = 1, max.cutoff = 3: These specify the minimum and maximum cutoff values for the expression levels of the gene to be included in the plot. Cells with expression levels outside this range will not be displayed.
By executing the a feature plot will be generated specifically for the "MS4A1" gene. Only cells with expression levels of "MS4A1" within the range of 1 to 3 will be displayed in the plot. This visualization allows for the focused examination of cells that exhibit expression levels of the "MS4A1" gene within the specified cutoff range, enabling the identification of specific cell subsets or clusters with moderate expression of the gene.

we don’t use this one use 14 instead

```{r make_seurat68}
FeaturePlot(pbmc3k.final, features = c("MS4A1", "PTPRCAP"), min.cutoff = "q10", max.cutoff = "q90")
```
FeaturePlot(): This function is used to create a feature plot, which displays the expression levels of genes of interest in individual cells or cell clusters.
features = c("MS4A1", "PTPRCAP"): This parameter specifies the genes of interest that will be plotted in the feature plot. In this case, the genes "MS4A1" and "PTPRCAP" are selected.
min.cutoff = "q10", max.cutoff = "q90": These parameters specify the minimum and maximum cutoff values for the expression levels of the genes to be included in the plot. The values are specified as quantiles, where "q10" represents the 10th percentile and "q90" represents the 90th percentile.
By executing the code a feature plot will be generated for the "MS4A1" and "PTPRCAP" genes. Only cells with expression levels of these genes within the 10th to 90th percentile range will be displayed in the plot. This visualization allows for the examination of cells that exhibit expression levels of "MS4A1" and "PTPRCAP" within a specific range defined by their quantiles. It helps identify cell subsets or clusters with expression levels of these genes that fall within the specified cutoff range.

```{r make_seurat69}
FeaturePlot(pbmc3k.final, features = c("MS4A1", "CD79A"), blend = TRUE)
```
FeaturePlot(): This function is used to create a feature plot, which displays the expression levels of genes of interest in individual cells or cell clusters.
features = c("MS4A1", "CD79A"): This parameter specifies the genes of interest that will be plotted in the feature plot. In this case, the genes "MS4A1" and "CD79A" are selected.
blend = TRUE: This parameter specifies whether to blend the expression values of the genes in the plot. When blend is set to TRUE, the expression values of the genes will be combined, providing a unified representation of their expression levels.
By executing the code FeaturePlot(pbmc3k.final, features = c("MS4A1", "CD79A"), blend = TRUE), a feature plot will be generated for the "MS4A1" and "CD79A" genes. The expression levels of these genes will be blended together, resulting in a unified representation of their expression levels in the plot. This visualization allows for the comparison and identification of cell subsets or clusters that exhibit similar expression patterns of the "MS4A1" and "CD79A" genes.


```{r make_seurat70}
VlnPlot(pbmc3k.final, features = "percent.mt", split.by = "groups")
```
VlnPlot(): This function is used to create a violin plot, which displays the distribution of a numeric feature across different groups or conditions.
pbmc3k.final: This is the Seurat object that contains the scRNA-seq data.
features = "percent.mt": This parameter specifies the feature of interest, in this case, "percent.mt". The "percent.mt" feature represents the percentage of mitochondrial genes in each cell, which can serve as an indicator of cell quality and potential mitochondrial contamination.
split.by = "groups": This parameter specifies the grouping variable by which the violin plot will be split. The grouping variable "groups" contains categorical information about different groups or conditions associated with the cells. The violin plot will display separate violins for each group, allowing for the comparison of the distribution of the "percent.mt" feature between different groups.
By executing the code VlnPlot(pbmc3k.final, features = "percent.mt", split.by = "groups"), a violin plot will be generated for the "percent.mt" feature. The plot will be split into separate violins based on the values of the "groups" variable, allowing for the comparison of the distribution of the "percent.mt" feature between different groups. This visualization helps identify differences or similarities in the percentage of mitochondrial genes across different groups or conditions in the dataset, which can provide insights into the quality and composition of the cells.



```{r make_seurat71}
DotPlot(pbmc3k.final, features = features, split.by = "groups") + RotatedAxis()
```
DotPlot(): This function is used to create a dot plot, which displays the expression levels of multiple features (genes) across different groups or conditions.
features = features: This parameter specifies the features of interest that will be plotted in the dot plot. In this case, the features variable contains a vector of gene names.
split.by = "groups": This parameter specifies the grouping variable by which the dot plot will be split. The grouping variable "groups" contains categorical information about different groups or conditions associated with the cells. The dot plot will display separate panels for each group, allowing for the comparison of gene expression between different groups.
RotatedAxis(): This function is used to rotate the axis labels in the dot plot to improve readability when displaying a large number of features.
By executing the code DotPlot(pbmc3k.final, features = features, split.by = "groups") + RotatedAxis(), a dot plot will be generated for the specified features. The plot will be split into separate panels based on the values of the "groups" variable, allowing for the comparison of gene expression between different groups. The rotated axis labels help accommodate a larger number of features on the plot without overlapping, making it easier to interpret the expression patterns across different groups.





```{r make_seurat72}
DimPlot(pbmc3k.final)
```

```{r make_seurat73}
pbmc3k.final.no.umap <- pbmc3k.final
```
pbmc3k.final.no.umap: This is the name given to the new Seurat object that will be created.
pbmc3k.final: This is the original Seurat object containing the scRNA-seq data.
By executing the code pbmc3k.final.no.umap <- pbmc3k.final, a new Seurat object named pbmc3k.final.no.umap is created. This new object contains the same data and metadata as the original pbmc3k.final object, allowing me to perform further analyses or modifications without affecting the original object. the intention is to create a copy of the pbmc3k.final object to work with separately, potentially for comparing different analyses.






```{r make_seurat74}
pbmc3k.final.no.umap[["umap"]] <- NULL
```
NULL: By assigning NULL to the "umap" slot, we are effectively removing the UMAP dimensional reduction results from the Seurat object.
By executing the code pbmc3k.final.no.umap[["umap"]] <- NULL, the UMAP coordinates stored in the "umap" slot of the pbmc3k.final.no.umap Seurat object are removed. This can be useful if you no longer need or want to work with the UMAP results and want to free up memory or simplify downstream analyses.

```{r make_seurat75}
DimPlot(pbmc3k.final.no.umap) + RotatedAxis()
```
DimPlot(): This function is used to create a plot of the cells in a reduced dimensional space.
pbmc3k.final.no.umap: This is the Seurat object that contains the scRNA-seq data. In this case, it is specifically named pbmc3k.final.no.umap, indicating that the UMAP dimensional reduction step was not performed on this object.
RotatedAxis(): This function is used to rotate the axis labels in the plot to improve readability when displaying a large number of cells or clusters.
By executing the code DimPlot(pbmc3k.final.no.umap) + RotatedAxis(), a plot will be generated to visualize the cells in a reduced dimensional space. The specific reduced dimensional space depends on the previous steps performed on the Seurat object, such as dimensionality reduction techniques like PCA or t-SNE. The RotatedAxis() function is then used to rotate the axis labels, which can be particularly useful when there are many cells or clusters being visualized, making it easier to interpret and read the labels.



```{r make_seurat76}
DoHeatmap(pbmc3k.final, features = VariableFeatures(pbmc3k.final)[1:100], cells = 1:500, size = 4,
          angle = 90) + NoLegend()

```
DoHeatmap(): This function is used to create a heatmap of the expression levels of features in a Seurat object.
features = VariableFeatures(pbmc3k.final)[1:100]: This specifies the features (genes) to be included in the heatmap. In this case, it selects the top 100 variable features based on the variability of their expression across cells.
cells = 1:500: This specifies the subset of cells to be included in the heatmap. It selects cells from index 1 to 500.
size = 4: This sets the size of the cells in the heatmap. A larger size (4) is chosen here.
angle = 90: This sets the angle of the column labels in the heatmap to 90 degrees, making them vertical.
NoLegend(): This function is used to exclude the legend from the heatmap plot.
By executing the code DoHeatmap(pbmc3k.final, features = VariableFeatures(pbmc3k.final)[1:100], cells = 1:500, size = 4, angle = 90) + NoLegend(), a heatmap will be generated, visualizing the expression levels of the top 100 variable features in a subset of cells from the pbmc3k.final Seurat object. The larger cell size and vertical column labels enhance the readability of the heatmap, and the legend is not included in the plot.


```{r make_seurat77}
baseplot <- DimPlot(pbmc3k.final, reduction = "umap")
```
DimPlot: This is a function in Seurat that generates a dimensionality plot.
reduction = "umap": This parameter specifies that the UMAP dimensionality reduction method should be used for visualization.
UMAP (Uniform Manifold Approximation and Projection) is a popular non-linear dimensionality reduction technique that is often used to visualize single-cell RNA sequencing (scRNA-seq) data.
By executing the a UMAP-based dimensionality plot is generated. The plot visualizes the cells in the pbmc3k.final Seurat object based on their UMAP coordinates. 


```{r make_seurat78}
baseplot + labs(title = "Clustering of 2,700 PBMCs")
```
baseplot: This is the UMAP-based dimensionality plot object created earlier using the DimPlot function.
+: The plus sign (+) is used in R to add additional elements or modifications to an existing plot.
labs(title = "Clustering of 2,700 PBMCs"): This is a function in ggplot2 (which is used by Seurat for plotting) that sets the plot title. The title argument is set to "Clustering of 2,700 PBMCs". the title "Clustering of 2,700 PBMCs" is added to the baseplot UMAP plot. This enhances the plot by providing a descriptive title that summarizes the content or purpose of the plot.
(it Just adds a title to the plot)

```{r make_seurat79}
baseplot + ggmin::theme_powerpoint()
```
The function applies a custom theme to the baseplot,  The custom theme is designed to resemble the default styling of PowerPoint presentations.
By using this code I  can modify the appearance of the UMAP plot to look like a PowerPoint slide. This includes changes to the font size, color palette, background, axis labels, and other visual elements to create a cleaner and more presentable plot.

```{r make_seurat80}
baseplot + DarkTheme()
```
it make it dark themed
```{r make_seurat81}
baseplot + FontSize(x.title = 20, y.title = 20) + NoLegend()
```
baseplot: This is the UMAP-based dimensionality plot object created earlier using the DimPlot function.
+: The plus sign (+) is used in R to add additional elements or modifications to an existing plot.FontSize(x.title = 20, y.title = 20): This modifies the font size of the x-axis and y-axis titles. The x.title and y.title arguments are set to 20, indicating a font size of 20 points for both axis titles.
+: Another plus sign (+) is used to add another modification to the plot.
NoLegend(): This removes the legend from the plot. The NoLegend() function is used to specify that the plot should not include a legend.
By executing this code the font size of the x-axis and y-axis titles in the baseplot UMAP plot is increased to 20 points, and the legend is removed. The legend in a plot provides information about the different elements or categories represented in the plot. It helps to interpret the colors, shapes, or sizes used to distinguish different groups or variables. The legend typically includes labels or symbols that correspond to the different categories or variables present in the plot.In the context of the UMAP plot, the legend might represent different cell types, clusters, or groups of cells. Each unique category or group is assigned a specific color or symbol in the plot, and the legend provides a key to understand which color or symbol corresponds to each category.By using the NoLegend() function, the code removes the legend from the UMAP plot. This can be useful when the legend is not necessary or when it may overlap with other elements in the plot, resulting in a cleaner and less cluttered visualization. 


```{r make_seurat82}
plot <- FeaturePlot(pbmc3k.final, features = "MS4A1")
```
plot: This is the name of the variable assigned to store the plot object.
FeaturePlot: This is a function in Seurat that creates a feature plot, which displays the expression levels of one or more genes across cells.
pbmc3k.final: This is the Seurat object that contains the single-cell RNA-seq data.
features = "MS4A1": This specifies the gene of interest to be plotted. In this case, the gene "MS4A1" is selected.
By executing this code  the expression levels of the gene "MS4A1" in the cells of the pbmc3k.final Seurat object are visualized in the plot. 


```{r make_seurat83}
HoverLocator(plot = plot, information = FetchData(pbmc3k.final, vars = c("ident", "PC_1", "nFeature_RNA")))
```
HoverLocator: This is a function in Seurat that adds a hover locator to a plot.
plot = plot: This specifies the plot object to which the hover locator should be added.
information = FetchData(pbmc3k.final, vars = c("ident", "PC_1", "nFeature_RNA")): This specifies the information to be displayed when hovering over data points in the plot. This function in Seurat allows you to interactively explore the data points in the plot by hovering your mouse over them. When I hover over a data point, it displays additional information associated with that point, such as the cell identity, the value of the first principal component (PC_1), and the number of RNA features detected (nFeature_RNA).as it allows us to quickly identify specific cells or patterns of interest and retrieve relevant information about them without cluttering the plot with labels or annotations.
By using HoverLocator in combination with the FetchData function, you can seamlessly integrate this interactive capability into your plot and enhance your data exploration and analysis experience.


```{r make_seurat84}
pbmc3k.final <- RenameIdents(pbmc3k.final, DC = "CD14+ Mono")
```
pbmc3k.final: This is the name of the Seurat object.
RenameIdents: This is a function in Seurat that is used to rename the cell identity labels in a Seurat object.
pbmc3k.final: This specifies the Seurat object on which the cell identity labels should be renamed.
DC = "CD14+ Mono": This specifies the new label "DC" that should replace the existing label "CD14+ Mono". By executing this code the cell identity label "CD14+ Mono" in the pbmc3k.final Seurat object is renamed to "DC". This allows for consistent and meaningful cell identity annotations in downstream analyses and visualizations.


```{r make_seurat85}
plot <- DimPlot(pbmc3k.final, reduction = "umap")
```
plot: This is the name of the variable where the resulting plot will be stored.
DimPlot: This is a function in Seurat used to create scatter plots of cells based on a specified dimensionality reduction.
pbmc3k.final: This specifies the Seurat object from which the cells will be plotted.
reduction = "umap": This parameter specifies the dimensionality reduction method to be used for the plot. In this case, "umap" indicates that the UMAP algorithm was used to reduce the dimensionality of the data.By executing the code plot <- DimPlot(pbmc3k.final, reduction = "umap"), a scatter plot of cells is generated using the UMAP dimensionality reduction, and the plot is stored in the variable named plot. The resulting plot can be further customized or visualized by using the plot variable in subsequent code.

```{r make_seurat86}
select.cells <- CellSelector(plot = plot)
```
select.cells: This is the name of the variable where the resulting cell selection information will be stored.
CellSelector: This is a function in Seurat that creates a user interface for cell selection based on a provided plot.
plot = plot: This parameter specifies the plot on which the cell selection interface will be based. In this case, the plot variable generated from the previous code is used.
By executing this code a user interface will be displayed that
allows you to interactively select cells on the scatter plot. You can use this interface to choose specific
cells of interest for downstream analyses or further exploration. The selected cell information will be stored in the select.cells variable for further use.

select some cells from the plot and hit done 


```{r make_seurat87}
head(select.cells)
```

```{r make_seurat88}
Idents(pbmc3k.final, cells = select.cells) <- "NewCells"
```
Idents: This is a function in Seurat that allows you to assign or modify cell identities in a Seurat object.
pbmc3k.final: This is the Seurat object to which you want to assign new cell identities.
cells = select.cells: This parameter specifies the cells to which the new identity will be assigned. Here, select.cells contains the information about the selected cells from the previous step.
"NewCells": This is the new identity label that will be assigned to the selected cells.
By executing the code Idents(pbmc3k.final, cells = select.cells) <- "NewCells", the selected cells in the pbmc3k.final Seurat object will be assigned the identity label "NewCells". This allows you to differentiate and analyze these cells separately from the rest of the cells in the dataset.

```{r make_seurat89}
newcells.markers <- FindMarkers(pbmc3k.final, ident.1 = "NewCells", ident.2 = "CD14+ Mono", min.diff.pct = 0.3,
                                only.pos = TRUE)


```
FindMarkers: This is a function in Seurat that performs differential expression analysis between two cell populations.
ident.1 = "NewCells": This parameter specifies the first cell population for which differential expression will be calculated. In this case, it is the "NewCells" population, which was assigned as the identity label in the previous step.
ident.2 = "CD14+ Mono": This parameter specifies the second cell population for which differential expression will be calculated. Here, it is the "CD14+ Mono" population.
min.diff.pct = 0.3: This parameter determines the minimum percentage difference required for a gene to be considered differentially expressed. In this case, a gene must have a percentage difference of at least 0.3 between the two populations to be considered differentially expressed.
only.pos = TRUE: This parameter specifies that only positively differentially expressed genes will be returned in the results. Genes that are upregulated in the "NewCells" population compared to the "CD14+ Mono" population will be included.
By executing the code newcells.markers <- FindMarkers(pbmc3k.final, ident.1 = "NewCells", ident.2 = "CD14+ Mono", min.diff.pct = 0.3, only.pos = TRUE), you will obtain a table (newcells.markers) that contains the top differentially expressed genes between the "NewCells" and "CD14+ Mono" populations. The head(newcells.markers) command is then used to display the first few rows of the resulting table.


```{r make_seurat90}
head(newcells.markers)
```

```{r make_seurat91}
plot <- DimPlot(pbmc3k.final, reduction = "pca") + NoLegend()
```
DimPlot: This is a function in Seurat that creates a scatter plot of cells based on a dimensionality reduction technique. In this case, the reduction parameter is set to "pca" to use PCA as the dimensionality reduction method.
reduction = "pca": This parameter specifies the type of dimensionality reduction to use. PCA (Principal Component Analysis) is a commonly used method to reduce the dimensionality of high-dimensional datasets while preserving the most important features.
+ NoLegend(): This code adds the NoLegend() function to the plot, which removes the legend from the plot. This is useful when you want to create a plot without showing the legend.
By executing the code plot I will obtain a scatter plot (plot) that visualizes the cells in the pbmc3k.final Seurat object using the PCA dimensionality reduction method. The NoLegend() function ensures that the legend is not displayed in the plot.

```{r make_seurat92}
LabelClusters(plot = plot, id = "ident")
```
LabelClusters: This function in Seurat is used to add cluster labels to a plot.
plot: This is the scatter plot (plot object) to which the cluster labels will be added.
id = "ident": This parameter specifies the identity column in the Seurat object that contains the cluster information. In this case, the cluster identity information is stored in the column named "ident" of the pbmc3k.final Seurat object.
By executing the code LabelClusters(plot = plot, id = "ident"), you will add cluster labels to the scatter plot plot based on the cluster information stored in the "ident" column of the pbmc3k.final Seurat object. The resulting plot will display the cluster labels, allowing you to identify the clusters visually.
in our case it will show each cell type

```{r make_seurat93}
LabelPoints(plot = plot, points = TopCells(object = pbmc3k.final[["pca"]]), repel = TRUE, max.overlaps = 100, xnudge = 0, ynudge = 0)
```
LabelPoints: This function in Seurat is used to add labels to specific points on a plot.
plot: This is the scatter plot (plot object) to which the labels will be added.
points = TopCells(object = pbmc3k.final[["pca"]]): This parameter specifies the points on the plot to which labels should be added. The TopCells() function is used to select the top cells based on some criteria. In this case, it selects the top cells based on the PCA scores obtained from the pbmc3k.final Seurat object.
repel = TRUE: This parameter determines whether the labels should be automatically adjusted to avoid overlapping with each other. By setting it to TRUE, the labels will be repelled from each other to improve readability.
By executing the code LabelPoints(plot = plot, points = TopCells(object = pbmc3k.final[["pca"]]), repel = TRUE), you will add labels to the top cells on the scatter plot plot based on the PCA scores obtained from the pbmc3k.final Seurat object. The resulting plot will display the labels for the top cells, allowing you to identify and visualize them more easily.


```{r make_seurat94}
plot5 <- DimPlot(pbmc3k.final)
```
The code plot5 <- DimPlot(pbmc3k.final) creates a scatter plot (plot5 object) using the DimPlot() function in Seurat. 
The DimPlot() function in Seurat is used to visualize cells in reduced dimensions. It takes the Seurat object pbmc3k.final as input and generates a scatter plot of the cells in the reduced dimension space. By default, the function uses the first two dimensions (UMAP_1 and UMAP_2) for visualization.
Executing the code will generate the scatter plot plot5 based on the default dimensionality reduction method in the pbmc3k.final Seurat object. You can further customize the plot by adding labels, color-coding cells based on specific metadata, or adjusting the appearance using additional Seurat functions and plot annotations.




```{r make_seurat95}
plot6 <- FeatureScatter(pbmc3k.final, feature1 = "LYZ", feature2 = "CCL5")
```
The code plot6 <- FeatureScatter(pbmc3k.final, feature1 = "LYZ", feature2 = "CCL5") creates a scatter plot (plot6 object) using the FeatureScatter() function in Seurat. The scatter plot is generated by plotting the expression levels of two features, "LYZ" and "CCL5", against each other.
The FeatureScatter() function in Seurat allows you to visualize the relationship between two features (genes) by plotting their expression levels for each cell in a scatter plot. In this case, the feature1 argument is set to "LYZ" and feature2 argument is set to "CCL5", which means the expression levels of the "LYZ" gene will be plotted on the x-axis and the expression levels of the "CCL5" gene will be plotted on the y-axis.
Executing the code  will generate the scatter plot plot6 showing the expression levels of "LYZ" and "CCL5" for each cell in the pbmc3k.final Seurat object. This plot can help identify any potential correlation or patterns between the expression levels of these two genes in the cells.


```{r make_seurat96}
plot5 + plot6
```

```{r make_seurat97}
(plot5 + plot6) & NoLegend()
```
The code (plot5 + plot6) & NoLegend() combines two plots, plot5 and plot6, and removes the legend from the combined plot.
In Seurat, the + operator is used to combine multiple plots or layers together. In this case, plot5 + plot6 combines the plots plot5 and plot6 into a single plot, stacking them vertically.
The & operator is used to modify the plot options. In this case, NoLegend() is applied to remove the legend from the combined plot.
Executing (plot5 + plot6) & NoLegend() will generate a new plot that combines plot5 and plot6, with no legend displayed. This can be useful for visualizing multiple plots together without the clutter of the legend.

